actionpack/lib/action_dispatch/routing/inspector.rb

```diff
@@ -33,11 +33,11 @@ def reqs
       end
 
       def controller
-        requirements[:controller] || ':controller'
+        parts.include?(:controller) ? ':controller' : requirements[:controller]
       end
 
       def action
-        requirements[:action] || ':action'
+        parts.include?(:action) ? ':action' : requirements[:action]
```

Comment:
¯ \ _ (ツ) _ / ¯  I don't remember what parts is right here, if it's a hash we could `parts.key?(:action)` if it's an array, then maybe when it's initialized we could loop over and pull out those to values `:controller` and `:action` so that we only have to iterate once.

The routing code is very performance sensitive. I'm fine with getting a solution on there then working to speed it back up if it's an issue.